app/routes/session.py

```diff
@@ -184,7 +186,46 @@ def _set_questionnaire_supplementary_data(
     _validate_supplementary_data_lists(
         supplementary_data=supplementary_data["data"], schema=schema
     )
-    questionnaire_store.set_supplementary_data(supplementary_data["data"])
+    _set_supplementary_data(
+        questionnaire_store=questionnaire_store,
+        schema=schema,
+        supplementary_data=supplementary_data["data"],
+    )
+
+
+def _set_supplementary_data(
+    *,
+    questionnaire_store: QuestionnaireStore,
+    schema: QuestionnaireSchema,
+    supplementary_data: dict,
+) -> None:
+    """
+    Adds the supplementary data to the questionnaire store which:
+    1) removes any old list items and answers
+    2) Updates block and section progress to reflect any newly unlocked questions due to supplementary data list changes
+    """
+    router = Router(
+        schema=schema,
+        answer_store=questionnaire_store.answer_store,
+        metadata=questionnaire_store.metadata,
+        list_store=questionnaire_store.list_store,
+        progress_store=questionnaire_store.progress_store,
+        response_metadata=questionnaire_store.response_metadata,
+        supplementary_data_store=questionnaire_store.supplementary_data_store,
+    )
+    questionnaire_store_updater = QuestionnaireStoreUpdater(
```

Comment:
I think I would lean toward having the non location questionnaire store updater class, rather than having the additional location checks, but curious to see what others think?